[Port] Headcrabs

PuroSlavKing opened the pull request:
<!-- Guidelines: https://docs.spacestation14.io/en/getting-started/pr-guideline -->

## About the PR
<!-- What did you change? -->
Taken from WWDP

## Why / Balance
<!-- Discuss how this would affect game balance or explain why it was changed. Link any relevant discussions or issues. -->
Thats fun

## Technical details
<!-- Summary of code changes for easier review. -->
Renamed changeling "headcrab" to "headslug", because that how this thing named

## Media
<!-- Attach media if the PR makes ingame changes (clothing, items, features, etc). 
Small fixes/refactors are exempt. Media may be used in SS14 progress reports with credit. -->

Mind transfer:

https://github.com/user-attachments/assets/77666aa0-f405-4708-91cc-ef9f43271312





## Requirements
<!-- Confirm the following by placing an X in the brackets [X]: -->
- [ ] I have read and am following the [Pull Request and Changelog Guidelines](https://docs.spacestation14.com/en/general-development/codebase-info/pull-request-guidelines.html).
- [x] I have added media to this PR or it does not require an ingame showcase.
<!-- You should understand that not following the above may get your PR closed at maintainer’s discretion -->

## Breaking changes
<!-- List any breaking changes, including namespaces, public class/method/field changes, prototype renames; and provide instructions for fixing them.
This will be posted in #codebase-changes. -->

**Changelog**
<!-- Add a Changelog entry to make players aware of new features or changes that could affect gameplay.
Make sure to read the guidelines and take this Changelog template out of the comment block in order for it to show up.
Changelog must have a :cl: symbol, so the bot recognizes the changes and adds them to the game's changelog. -->
<!--
:cl:
- add: Added fun!
- remove: Removed fun!
- tweak: Changed fun!
- fix: Fixed fun!
-->

Aviu00 commented on Content.Server/_White/Headcrab/HeadcrabComponent.cs:
```diff
@@ -0,0 +1,38 @@
+    /// <summary>
+    /// WorldTargetAction
+    /// </summary>
+    [DataField(required: true, customTypeSerializer: typeof(PrototypeIdSerializer<EntityPrototype>))]
+    public string JumpAction = "ActionHeadcrabJump";
```
```suggestion
    [DataField]
    public EntProtoId JumpAction = "ActionHeadcrabJump";
```
Use `EntProtoId` instead of searializer. Also this doesn't need to be `(required: true)` since you already specify the action.

Aviu00 commented on Content.Server/_White/Headcrab/HeadcrabComponent.cs:
```diff
@@ -0,0 +1,38 @@
+    public string JumpAction = "ActionHeadcrabJump";
+
+    [DataField]
+    public float ParalyzeTime = 3f;
```
```suggestion
    public TimeSpan ParalyzeTime = TimeSpan.FromSeconds(3);
```
Use `TimeSpan` instead of float.

Aviu00 commented on Content.Server/_White/Headcrab/HeadcrabComponent.cs:
```diff
@@ -0,0 +1,38 @@
+    public float ParalyzeTime = 3f;
+
+    [DataField]
+    public int ChancePounce = 33;
```
```suggestion
    public float ChancePounce = 0.33f;
```
Use float for random chance. To calculate the probability, do `_random.Prob(ChancePounce)`.

Aviu00 commented on Content.Server/_White/Headcrab/HeadcrabSystem.cs:
```diff
@@ -0,0 +1,202 @@
+        }
+    }
+
+        private bool TryEquipHeadcrab(EntityUid uid, EntityUid target, HeadcrabComponent component)
+    {
```
```suggestion
    private bool TryEquipHeadcrab(EntityUid uid, EntityUid target, HeadcrabComponent component)
    {
```
Formatting.

Aviu00 commented on Content.Server/_White/Headcrab/HeadcrabSystem.cs:
```diff
@@ -0,0 +1,202 @@
+
+        _inventory.TryGetSlotEntity(target, "head", out var headItem);
+        return !HasComp<IngestionBlockerComponent>(headItem)
+            && !_inventory.TryEquip(target, uid, "mask", true);
```
```suggestion
            && _inventory.TryEquip(target, uid, "mask", true);
```
Why is it returning false if we successfully equip the headcrab?

Aviu00 commented on Content.Server/_White/Headcrab/HeadcrabSystem.cs:
```diff
@@ -0,0 +1,202 @@
+    private void OnMeleeHit(EntityUid uid, HeadcrabComponent component, MeleeHitEvent args)
+    {
+        if (!args.HitEntities.Any()
+            || _random.Next(1, 101) <= component.ChancePounce)
```
Use `_random.Prob()`

Aviu00 commented on Content.Server/_White/Headcrab/HeadcrabSystem.cs:
```diff
@@ -0,0 +1,202 @@
+        var direction = mapCoords.Position - _transform.GetMapCoordinates(xform).Position;
+
+        _throwing.TryThrow(uid, direction, 7F, uid, 10F);
+        if (component.JumpSound != null)
+            _audioSystem.PlayPvs(component.JumpSound, uid, component.JumpSound.Params);
```
```suggestion
        _audioSystem.PlayPvs(component.JumpSound, uid, component.JumpSound.Params);
```
`PlayPvs` checks if sound is null so you don't have to/

Aviu00 commented on Content.Server/_White/Headcrab/HeadcrabSystem.cs:
```diff
@@ -0,0 +1,202 @@
+        var mapCoords = _transform.ToMapCoordinates(args.Target);
+        var direction = mapCoords.Position - _transform.GetMapCoordinates(xform).Position;
+
+        _throwing.TryThrow(uid, direction, 7F, uid, 10F);
```
You should probably check if direction is zero vector, normalize it and multiply it by some value so headcrabs don't fly too far if they click far away.

Aviu00 commented on Resources/Prototypes/_White/Entities/Mobs/NPCs/headcrab.yml:
```diff
@@ -0,0 +1,145 @@
+    quickEquip: false
+    slots:
+      - MASK
+      - HEAD
```
Why can it be equipped on head? Isn't it only supposed to be attached to mask slot?

PuroSlavKing replied:
> Why can it be equipped on head? Isn't it only supposed to be attached to mask slot?

Because you can pick it up when its dead, why you cant put it on your head?

Aviu00 commented on Content.Server/_White/Headcrab/HeadcrabSystem.cs:
```diff
@@ -0,0 +1,202 @@
+        RemCompDeferred<PacifiedComponent>(uid);
+        var combatMode = EnsureComp<CombatModeComponent>(uid);
+        _combat.SetInCombatMode(uid, true, combatMode);
+        EnsureComp<NPCMeleeCombatComponent>(uid);
```
You can't just remove and add `NPCMeleeCombatComponent` since it's managed by HTN operators. But if it works, then it's fine I guess.

Aviu00 commented on Resources/Prototypes/_White/Entities/Mobs/NPCs/headcrab.yml:
```diff
@@ -0,0 +1,145 @@
+      - MASK
+      - HEAD
+  - type: Physics
+    bodyType: Dynamic
```
Why is it `Dynamic`? Should be `KinematicController`.

Aviu00 reviewed (changes requested):
There should be a way for headcrab to detach itself from zombie so it's non stuck forever. Or even better, if it kills someone, turn them into zombie and let headcrab assume control of them, detach headcrab when the zombie dies.

Aviu00 reviewed (changes requested):
Nice, though here are things that need to be changed:
1. headcrab_jump.ogg should be mono.
2. Headcrabs should be detached when a person crits, not dies.
3. It would be cooler if headcrab's mind (if it has one) would transfer to the victim when it attaches itself and maybe shooting should be blocked in this case. Also headcrab ghost role should be disabled if it's attached to someone.
4. Instead of ghosting, victim's mind should be stored in some entity, maybe spawn invisible entity and parent it to the victim's entity or do some container shenanigans.
5. Headcrab's eating head damage maybe should be increased, so it can't assume control for too long.
6. If headcrab is attached to an entity and round restarts, it errors in console and fails to start a round, probably need to check if entity is terminating on unequipping.
7. Detach headcrab on polymorph or entity termination.

Aviu00 commented:
Also instantly assuming control is probably a bad idea. It should happen if the victim below a certain health threshold or something like that so they can actually struggle. And after detaching, headcrab shouldn't be able to jump/attach by hitting for some time.

PuroSlavKing commented on Resources/Prototypes/_White/Entities/Mobs/NPCs/headcrab.yml:
```diff
@@ -0,0 +1,145 @@
+
+# HEADCRABS
+
+- type: entity
+  parent: BaseMobHeadcrab
+  id: MobHeadcrab
+  components:
+  - type: Sprite
+    sprite: _White/Mobs/Animals/headcrab.rsi
+  - type: Clothing
+    sprite: _White/Mobs/Animals/headcrab.rsi
+
```
thats not headslug!!! STOP REVIEW LIKE THAT!!!

PuroSlavKing commented:
> Nice, though here are things that need to be changed:
> 
> 1. headcrab_jump.ogg should be mono.
> 2. Headcrabs should be detached when a person crits, not dies.
> 3. It would be cooler if headcrab's mind (if it has one) would transfer to the victim when it attaches itself and maybe shooting should be blocked in this case. Also headcrab ghost role should be disabled if it's attached to someone.
> 4. Instead of ghosting, victim's mind should be stored in some entity, maybe spawn invisible entity and parent it to the victim's entity or do some container shenanigans.
> 5. Headcrab's eating head damage maybe should be increased, so it can't assume control for too long.
> 6. If headcrab is attached to an entity and round restarts, it errors in console and fails to start a round, probably need to check if entity is terminating on unequipping.
> 7. Detach headcrab on polymorph or entity termination.

1. Already done
2. Thats how its work, dead states used for headcrab, not host
3. Nuh.
4. Done, updated video included
5. Done
6. Fixed?
7. Skill issue, idk how do this

Aviu00 commented on Content.Server/_White/Headcrab/HeadcrabSystem.cs:
```diff
@@ -0,0 +1,294 @@
+        var headcrabHasMind = _mindSystem.TryGetMind(uid, out var mindId, out var mind);
+        var hostHasMind = _mindSystem.TryGetMind(args.Equipee, out var hostMindId, out var hostMind);
+
+        if (!headcrabHasMind && !hostHasMind)
+                return;
```
```suggestion
        if (!headcrabHasMind && !hostHasMind)
             return;
```
Formatting.

Aviu00 commented on Resources/Prototypes/_White/Entities/Mobs/NPCs/headcrab.yml:
```diff
@@ -0,0 +1,145 @@
+    description: ghost-role-information-headcrab-description 
+    raffle:
+      settings: short
+  - type: Sprite
```
```suggestion
  - type: Sprite
    offset: 0, 0.2
```
Add some Y offset to the sprite, Their actual sprite is below their fixture so they can "phase" themselves into a wall to the south of them.

Aviu00 commented on Content.Server/_White/Headcrab/HeadcrabSystem.cs:
```diff
@@ -0,0 +1,294 @@
+            _mindSystem.TransferTo(hostMindId, uid, mind: hostMind);
+        }
+
+        _action.AddAction(uid, ref component.JumpActionEntity, component.JumpAction, uid);
```
This just adds the second action.

Aviu00 reviewed (changes requested):
If, before attaching, headcrab has mind and victim doesn't, headcrab's mind will be attached to victim's body after it's unequipped.
/suicide should be blocked when headcrab is controlling victim's body to prevent round removals.
Headcrabs can't attach themselves if a person is wearing any mask, so they are really easy to counter, is it intended?

Aviu00 reviewed (changes requested):
If headcrab has mind and victim doesn't, after unequipping, headcrab would be controlled by both npc and player at the same time for some reason. If victim has mind and headcrab doesn't, after detaching, victim's mind would be transferred to headcrab and it would be controlled by both npc and player. Weird stuff.

PuroSlavKing commented:
> If headcrab has mind and victim doesn't, after unequipping, headcrab would be controlled by both npc and player at the same time for some reason. If victim has mind and headcrab doesn't, after detaching, victim's mind would be transferred to headcrab and it would be controlled by both npc and player. Weird stuff.

goddamn, may i just close this PR??

Aviu00 commented:
> goddamn, may i just close this PR?? 😭

Do as you wish.
